### ISSUE TITLE
Invalid aggregation

### ISSUE THREAD

**Issue by HEmile** [lucadiliello/semantic-loss-pytorch#1]
Hi there! 
I think this loss function does not aggregate the loss correctly. It does
```
        wmc = torch.mean(wmc_per_sample)
        loss = -torch.log(wmc)
```
Ie: The logarithm of the mean wmc per sample.

But I'm pretty sure the Semantic Loss computes the average log-wmc, ie
```
        log_wmc = torch.log(wmc_per_sample)
        loss = -log_wmc.mean()
```
Is this correct?

**Comment by lucadiliello** [comment]
Hi @HEmile, thanks for your interest in our project. You are right. However, as shown in `Equation 7` of our paper, it looks that by doing the average first and the log last the training is more stable.

Here you can find our [paper](https://arxiv.org/pdf/2007.13197.pdf) with the explanation.

**Comment by HEmile** [comment]
Correct me if I'm wrong! But isn't this something different? I mean the aggregation over the batch dimension (ie, the different samples), not the logarithm of whether a particular x evaluates to true.

**Comment by lucadiliello** [comment]
How would you write the computation then? Because by doing:
```python
log_wmc = torch.log(wmc_per_sample)
loss = -log_wmc.mean()
```
you risk that samples that don't satisfy the constraints (especially in the early steps of training) evaluate to very very small values. This was an even larger problem with `fp16` in our experiments.

**Comment by HEmile** [comment]
I see. That's why I thought you did it (but I don't think this is easy to see from the equation you mentioned, haha). 

Your reasoning makes a lot of sense! But the downside is that it cannot 'scale' the gradients to low-probability samples as much if I understand correctly (ie: Low-probability and high-probability losses will be treated equivalently. I'll try both, and see if it matters! Thanks :)
